presto-hive/src/main/java/com/facebook/presto/hive/statistics/ParquetQuickStatsBuilder.java

```diff
@@ -0,0 +1,366 @@
+            filesCount++;
+            Path path = file.getPath();
+            long fileSize = file.getLength();
+
+            HiveFileContext hiveFileContext = new HiveFileContext(
+                    true,
+                    NO_CACHE_CONSTRAINTS,
+                    Optional.empty(),
+                    OptionalLong.of(fileSize),
+                    OptionalLong.empty(),
+                    OptionalLong.empty(),
+                    file.getFileModifiedTime(),
+                    false);
+
+            HdfsContext hdfsContext = new HdfsContext(session, table.getSchemaName(), table.getTableName());
+            Configuration configuration = hdfsEnvironment.getConfiguration(hdfsContext, path);
+
+            footerFetchCompletableFutures.add(supplyAsync(() -> {
+                try (FSDataInputStream inputStream = hdfsEnvironment.getFileSystem(hdfsContext, path).openFile(path, hiveFileContext);
+                        ParquetDataSource parquetDataSource = buildHdfsParquetDataSource(inputStream, path, stats)) {
+                    ParquetFileMetadata parquetFileMetadata = readFooter(parquetDataSource,
+                            fileSize,
+                            createDecryptor(configuration, path),
+                            getReadNullMaskedParquetEncryptedValue(session));
+
+                    return parquetFileMetadata.getParquetMetadata();
+                }
+                catch (Exception e) {
+                    log.error(e);
+                    throw new RuntimeException(e);
+                }
+            }, footerFetchExecutor));
+        }
+
+        // Record a metric about how many files were seen
+        session.getRuntimeStats().addMetricValue(String.format("ParquetQuickStatsBuilder/FileCount/%s/%s", table.getTableName(), partitionId), RuntimeUnit.NONE, filesCount);
+
+        HashMap<ColumnPath, ColumnQuickStats<?>> rolledUpColStats = new HashMap<>();
+        try {
+            // Wait for footer reads to finish
+            CompletableFuture<Void> overallCompletableFuture = CompletableFuture.allOf(footerFetchCompletableFutures.toArray(new CompletableFuture[0]));
+            overallCompletableFuture.get(footerFetchTimeoutMillis, MILLISECONDS);
+
+            for (CompletableFuture<ParquetMetadata> future : footerFetchCompletableFutures) {
+                ParquetMetadata parquetMetadata = future.get();
+                processColumnMetadata(parquetMetadata, rolledUpColStats);
+            }
+        }
+        catch (InterruptedException | ExecutionException | TimeoutException e) {
+            log.error(e, "Failed to read/build stats from parquet footer");
+            throw new RuntimeException(e);
+        }
+
+        if (rolledUpColStats.isEmpty()) {
+            return PartitionQuickStats.EMPTY;
+        }
+        return new PartitionQuickStats(partitionId, rolledUpColStats.values(), filesCount);
+    }
+
+    @Managed
+    @Nested
+    public ThreadPoolExecutorMBean getExecutor()
```

Comment:
Should we add additional stats, such as how long it took to open a footer, how many bytes were read, and simple call count metrics like how many times quick stats were requested and how many files were opened?

Reply:
I have simple call count metrics in the QuickStatsProvider (as JMX)-
 
https://github.com/aaneja/presto/blob/39f417eebe6b4a7d2912efb97589cbee634762cd/presto-hive/src/main/java/com/facebook/presto/hive/statistics/QuickStatsProvider.java#L120-L136

The latency to build Quick stats via a provider (as runtime stats) which are a proxy for the worst case footer fetch latency (since we fetch footers concurrently)-

https://github.com/aaneja/presto/blob/39f417eebe6b4a7d2912efb97589cbee634762cd/presto-hive/src/main/java/com/facebook/presto/hive/statistics/QuickStatsProvider.java#L351

Number of footers that were scanned for a partition  (as runtime stats) -

https://github.com/aaneja/presto/blob/39f417eebe6b4a7d2912efb97589cbee634762cd/presto-hive/src/main/java/com/facebook/presto/hive/statistics/ParquetQuickStatsBuilder.java#L323-L324

I can track total bytes fetched, but since parquet footers are usually small (expected footer size 16K bytes) and fetched in a single FS call, we can estimate them via `file_count*16K`. FS stats from the coordinator (like `PrestoS3FileSystemStats`) should also give us a detailed picture for debugging

Reply:
I've added some more JMX metrics - 

https://github.com/aaneja/presto/blob/90fcf6772af5612f7d798b3a28b857233ceacb1e/presto-hive/src/main/java/com/facebook/presto/hive/statistics/ParquetQuickStatsBuilder.java#L256-L275

https://github.com/aaneja/presto/blob/90fcf6772af5612f7d798b3a28b857233ceacb1e/presto-hive/src/main/java/com/facebook/presto/hive/statistics/QuickStatsProvider.java#L140-L145